include/swift/SIL/MemAccessUtils.h

```diff
@@ -0,0 +1,294 @@
+    case Global:
+      return true;
+    case Class:
+    case Argument:
+    case Nested:
+    case Unidentified:
+      return false;
+    }
+  }
+  
+  bool isDistinct(const AccessedStorage &other) const {
+    return isUniquelyIdentified() && other.isUniquelyIdentified()
+      && (*this != other);
+  }
+
+  /// Returns the ValueDecl for the underlying storage, if it can be
+  /// determined. Otherwise returns null. For diagnostics and determining
+  /// `let` variables.
+  const ValueDecl *getDecl(SILFunction *F) const;
+};
+} // end namespace swift
+
+namespace llvm {
+
+/// Enable using AccessedStorage as a key in DenseMap.
+template <> struct DenseMapInfo<swift::AccessedStorage> {
+  static swift::AccessedStorage getEmptyKey() {
+    return swift::AccessedStorage(swift::SILValue::getFromOpaqueValue(
+                                    llvm::DenseMapInfo<void *>::getEmptyKey()),
+                                  swift::AccessedStorage::Unidentified);
+  }
+
+  static swift::AccessedStorage getTombstoneKey() {
+    return swift::AccessedStorage(
+      swift::SILValue::getFromOpaqueValue(
+        llvm::DenseMapInfo<void *>::getTombstoneKey()),
+      swift::AccessedStorage::Unidentified);
+  }
+
+  static unsigned getHashValue(swift::AccessedStorage storage) {
+    switch (storage.getKind()) {
+    case swift::AccessedStorage::Box:
+    case swift::AccessedStorage::Stack:
+    case swift::AccessedStorage::Nested:
+    case swift::AccessedStorage::Unidentified:
+      return DenseMapInfo<swift::SILValue>::getHashValue(storage.getValue());
+    case swift::AccessedStorage::Argument:
+      return storage.getParamIndex();
+    case swift::AccessedStorage::Global:
+      return DenseMapInfo<void *>::getHashValue(storage.getGlobal());
+    case swift::AccessedStorage::Class: {
+      const swift::ObjectProjection &P = storage.getObjectProjection();
+      return llvm::hash_combine(P.getObject(), P.getProjection());
+    }
+    }
+    llvm_unreachable("Unhandled AccessedStorageKind");
+  }
+
+  static bool isEqual(swift::AccessedStorage LHS, swift::AccessedStorage RHS) {
+    if (LHS.getKind() != RHS.getKind())
+      return false;
+
+    switch (LHS.getKind()) {
+    case swift::AccessedStorage::Box:
+    case swift::AccessedStorage::Stack:
+    case swift::AccessedStorage::Nested:
+    case swift::AccessedStorage::Unidentified:
+      return LHS.getValue() == RHS.getValue();
+    case swift::AccessedStorage::Argument:
+      return LHS.getParamIndex() == RHS.getParamIndex();
+    case swift::AccessedStorage::Global:
+      return LHS.getGlobal() == RHS.getGlobal();
+    case swift::AccessedStorage::Class:
+        return LHS.getObjectProjection() == RHS.getObjectProjection();
+    }
+    llvm_unreachable("Unhandled AccessedStorageKind");
+  }
+};
+
+} // end namespace llvm
+
+namespace swift {
+
+/// Given an address accessed by an instruction that reads or modifies
+/// memory, return an AccessedStorage that identifies the formal access.
```

Comment:
Improved comment #1:
```
/// Represents the identity of a storage location being accessed.
///
/// AccessedStorage may be one of several kinds of "identified" storage
/// locations, or may be valid, but Unidentified storage. An identified location
/// is known to identify the base of the accessed storage, whether that is a
/// SILValue that produces the base address, or a variable
/// declaration. "Uniquely identified" storage refers to identified storage that
/// cannot be aliased. For example, local allocations are uniquely identified,
/// while global variables and class properties are not. Unidentified storage is
/// associated with a SILValue that produces the accessed address but has not
/// been determined to be the base of a storage location. It may, for example,
/// be a SILPHIArgument.
///
/// An invalid AccessedStorage object is marked Unidentified and contains an
/// invalid value. This signals that analysis has failed to recognize an
/// expected address producer pattern. Over time, more aggressive
/// SILVerification could allow the optimizer to aggressively assert that
/// AccessedStorage is always valid.
///
/// Note that the SILValue that represents a storage location is not
/// necessarilly an address type. It may instead be a SILBoxType.
///
/// AccessedStorage hashing and comparison is used to determine when two
/// 'begin_access' instructions access the same or disjoint underlying
/// locations.
///
/// Equality guarantees that two AccessStorage values refer to the same
/// memory if both values are valid.
///
/// Inequality does not guarantee that two identified AccessStorage values
/// distinct. Inequality does guarantee that two *uniquely* identified
/// AccessStorage values are distinct.
class AccessedStorage {
```
Improved comment #2:
```
/// Given an address accessed by an instruction that reads or modifies
/// memory, return an AccessedStorage object that identifies the formal access.
///
/// The returned AccessedStorage represents the best attempt to find the base of
/// the storage location being accessed at `sourceAddr`. This may be a fully
/// identified storage base of known kind, or a valid but Unidentified storage
/// location, such as a SILPHIArgument.
///
/// This may return an invalid storage object if the address producer is not
/// recognized by a whitelist of recognizable access patterns. The result must
/// always be valid when `sourceAddr` is used for formal memory access, i.e. as
/// the operand of begin_access.
///
/// If `sourceAddr` is produced by a begin_access, this returns a Nested
/// AccessedStorage kind. This is useful for exclusivity checking to distinguish
/// between a nested access vs. a conflict.
AccessedStorage findAccessedStorage(SILValue sourceAddr);
```
Improved comment #3:
```
/// Given an address accessed by an instruction that reads or modifies
/// memory, return an AccessedStorage that identifies the formal access, looking
/// through any Nested access to find the original storage.
///
/// This is identical to findAccessedStorage(), but never returns Nested
/// storage.
AccessedStorage findAccessedStorageOrigin(SILValue sourceAddr);
```